File: marketdna-backend/app/models/quant_strategies.py
"""Pydantic models for the four quantitative strategy modules."""
from pydantic import BaseModel


# ─── Cross-Sectional Momentum ─────────────────────────────────────────────────

class MomentumItem(BaseModel):
    symbol: str
    sector: str
    ret_12_1m: float       # 12-1 month skip-one momentum return %
    ret_3m: float
    ret_1m: float
    rank: int              # 1 = strongest momentum
    quintile: int          # 1 = top 20%, 5 = bottom 20%
    signal: str            # "Strong Long" | "Long" | "Neutral" | "Avoid" | "Strong Avoid"
    options_strategy: str

class MomentumResult(BaseModel):
    items: list[MomentumItem]
    top_5: list[str]
    bottom_5: list[str]
    as_of: str


# ─── Mean Reversion Z-Score ───────────────────────────────────────────────────

class MeanReversionItem(BaseModel):
    symbol: str
    sector: str
    z_score_20d: float     # (close - SMA20) / std20
    z_score_50d: float
    bb_pct_b: float        # Bollinger Band %B  0=lower, 1=upper
    rsi: float
    reversion_score: float # composite −100 to +100  (positive = overbought)
    signal: str            # "Overbought" | "Near OB" | "Neutral" | "Near OS" | "Oversold"
    options_strategy: str

class MeanReversionResult(BaseModel):
    items: list[MeanReversionItem]
    overbought: list[str]  # signal == "Overbought"
    oversold: list[str]    # signal == "Oversold"
    as_of: str


# ─── Realized Volatility Rank ─────────────────────────────────────────────────

class VolRankItem(BaseModel):
    symbol: str
    sector: str
    hv20: float            # current annualised 20-day HV %
    hv20_1m_ago: float     # HV20 one month ago
    hv_rank: float         # percentile 0–100 vs trailing 252-day HV20 history
    vol_trend: str         # "Rising" | "Falling" | "Stable"
    regime: str            # "Very Low" | "Low" | "Normal" | "Elevated" | "High"
    options_strategy: str
    iv_action: str         # "Sell" | "Buy" | "Neutral"

class VolRankResult(BaseModel):
    items: list[VolRankItem]
    sell_candidates: list[str]   # hv_rank ≥ 70
    buy_candidates: list[str]    # hv_rank ≤ 30
    as_of: str


# ─── Sector Relative Strength Rotation ───────────────────────────────────────

class SectorItem(BaseModel):
    sector: str
    symbols: list[str]
    n_stocks: int
    ret_1m: float          # equal-weighted sector 1M return %
    ret_3m: float
    rs_1m: float           # sector ret_1m − NIFTY ret_1m
    rs_3m: float
    rank_3m: int           # 1 = strongest sector
    phase: str             # "Leading" | "Weakening" | "Lagging" | "Improving"

class SectorRotationResult(BaseModel):
    sectors: list[SectorItem]
    nifty_ret_1m: float
    nifty_ret_3m: float
    leading: list[str]
    improving: list[str]
    weakening: list[str]
    lagging: list[str]
    as_of: str


# ─── Combined scan result ─────────────────────────────────────────────────────

class QuantScanResult(BaseModel):
    momentum: MomentumResult
    mean_reversion: MeanReversionResult
    vol_rank: VolRankResult
    sector_rotation: SectorRotationResult
    computed_at: str

File: marketdna-web/src/types/quant_strategies.ts
export interface MomentumItem {
  symbol: string
  sector: string
  ret_12_1m: number
  ret_3m: number
  ret_1m: number
  rank: number
  quintile: number        // 1=top, 5=bottom
  signal: string
  options_strategy: string
}

export interface MomentumResult {
  items: MomentumItem[]
  top_5: string[]
  bottom_5: string[]
  as_of: string
}

export interface MeanReversionItem {
  symbol: string
  sector: string
  z_score_20d: number
  z_score_50d: number
  bb_pct_b: number
  rsi: number
  reversion_score: number   // positive = overbought
  signal: string
  options_strategy: string
}

export interface MeanReversionResult {
  items: MeanReversionItem[]
  overbought: string[]
  oversold: string[]
  as_of: string
}

export interface VolRankItem {
  symbol: string
  sector: string
  hv20: number
  hv20_1m_ago: number
  hv_rank: number
  vol_trend: string
  regime: string
  options_strategy: string
  iv_action: string
}

export interface VolRankResult {
  items: VolRankItem[]
  sell_candidates: string[]
  buy_candidates: string[]
  as_of: string
}

export interface SectorItem {
  sector: string
  symbols: string[]
  n_stocks: number
  ret_1m: number
  ret_3m: number
  rs_1m: number
  rs_3m: number
  rank_3m: number
  phase: string   // "Leading" | "Weakening" | "Lagging" | "Improving"
}

export interface SectorRotationResult {
  sectors: SectorItem[]
  nifty_ret_1m: number
  nifty_ret_3m: number
  leading: string[]
  improving: string[]
  weakening: string[]
  lagging: string[]
  as_of: string
}

export interface QuantScanResult {
  momentum: MomentumResult
  mean_reversion: MeanReversionResult
  vol_rank: VolRankResult
  sector_rotation: SectorRotationResult
  computed_at: string
}

// Display helpers
export const QUINTILE_COLOR: Record<number, string> = {
  1: '#22c55e',
  2: '#86efac',
  3: '#94a3b8',
  4: '#f87171',
  5: '#ef4444',
}

export const SIGNAL_COLOR: Record<string, string> = {
  'Strong Long':  '#22c55e',
  'Long':         '#86efac',
  'Neutral':      '#94a3b8',
  'Avoid':        '#f87171',
  'Strong Avoid': '#ef4444',
  'Overbought':   '#ef4444',
  'Near OB':      '#f87171',
  'Near OS':      '#86efac',
  'Oversold':     '#22c55e',
}

export const PHASE_COLOR: Record<string, string> = {
  'Leading':   '#22c55e',
  'Improving': '#86efac',
  'Weakening': '#f59e0b',
  'Lagging':   '#ef4444',
}

export const VOL_REGIME_COLOR: Record<string, string> = {
  'Very High': '#ef4444',
  'Elevated':  '#f87171',
  'Normal':    '#94a3b8',
  'Low':       '#86efac',
  'Very Low':  '#22c55e',
}
